### ISSUE TITLE
Crash when toggling visualizer

### ISSUE THREAD

**Issue by tranzystorek-io** [LeoRiether/tori#7]
**Describe the bug**

**To Reproduce**
1.  In `tori`, press the "Toggle visualizer" key

**Expected behavior**

The visualizer is enabled successfully

**Actual behavior**

`tori` crashes with:

```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: UnexpectedEof, message: "failed to fill whole buffer" }', src/visualizer/mod.rs:79:49
```

**Screenshots**

**System Information (please complete the following information):**
 - OS: Void Linux, `Linux vmware 6.1.28_1 #1 SMP PREEMPT_DYNAMIC Sat May 13 16:50:36 UTC 2023 x86_64 GNU/Linux`
 - tori version 0.1.2

**Additional context**

Reproduces on a vmware VM, with the following software:

WM: wayland/labwc 0.6.3
terminal: alacritty 0.12.0, zellij 0.36.0
mpv 0.35.1
cava 0.8.3

**Comment by LeoRiether** [comment]
Hi again o/

My guess is that cava opens fine, but instead of outputting the visualizer data it prints an error and exits, filling the buffer with that error, which we can't see now :upside_down_face:. What happens if you run `cava` in the terminal? Does it show a spectrum visualizer correctly?

I've added a ~~hopefully~~ better error message in 086ed20 too, so if the above doesn't work we'll be able to see exactly what cava wrote.

**Comment by tranzystorek-io** [comment]
Checked and my `cava` indeed fails because I'm using pipewire which isn't supported in 0.8.3 yet.

Unfortunately the reported process stdout isn't particularly interesting:

```
'Failed to read from the visualizer process. The received error was: failed to fill whole buffer. Process output: '
```

**Comment by tranzystorek-io** [comment]
@LeoRiether curious what you think, but IMO there's potentially going to be more users in a similar situation to mine, using pipewire and possibly having an old version of `cava` that doesn't support pipewire, so probably something more resilient than a panic here would be a good idea

**Comment by LeoRiether** [comment]
So I replaced the panic by an error notification that doesn't crash the program in a877e52 :)
Still want to replace cava later though.

![cava err](https://github.com/LeoRiether/tori/assets/8211902/0f87b82b-90be-44eb-89a7-627b717bf084)

**Comment by tranzystorek-io** [comment]
Tested that commit, LGTM

**Comment by tranzystorek-io** [comment]
> Still want to replace cava later though.

I did some research and found out https://github.com/karlstav/cava/issues/487 - cava is being made into a separate library. I'd be tempted to try making a rust wrapper for it as a learning experience and a way to help tori along the line :3

**Comment by tranzystorek-io** [comment]
I spent the last few days on writing the `cava`, and `cava-sys` libraries, but so far I only have a rough wrap around the `cavacore` library, as presented here: https://github.com/LukashonakV/cava/blob/master/include/cava/cavacore.h, not sure how useful that would be for tori, we currently seem to rely strongly on the `cava` binary being able to autoattach to the system audio daemon if I understand correctly.

**Comment by LeoRiether** [comment]
Oh, I expected cavacore to also handle the sampling of system audio. Yes, this would have to be implemented in tori itself (or in a library) for cavacore to be useful.

If you want to take a stab at implementing this, maybe you'll find the [audio_visualizer](https://github.com/phip1611/audio-visualizer) crate useful. It has a system audio recording module ([live_input.rs](https://github.com/phip1611/audio-visualizer/blob/main/src/dynamic/live_input.rs)). Technically we could get rid of cava entirely by using this crate, but if you want to use cavacore to generate the spectrum and something like `live_input` to get system audio, that's fine by me too :+1:

**Comment by tranzystorek-io** [comment]
cava seems to be better suited for players, with the noise reduction, normalising output etc.

**Comment by tranzystorek-io** [comment]
Been trying cpal and the live_input, but it seems to have a hard time detecting an alsa hw:Loopback device for me to capture from :/

**Comment by LeoRiether** [comment]
Great job! The demo is pretty cool :)

Does this, even if just in theory, also work in Windows and Mac? ALSA only exists in Linux, so you'd have to use something else to capture audio, right? After #14 is merged tori will finally build on these platforms, so it would be nice if the visualizer worked there too!

**Comment by tranzystorek-io** [comment]
> Does this, even if just in theory, also work in Windows and Mac?

The core of this visualizer is just a `ratatui` widget and the libcava wrapper I've written, but the sound input is pretty much only tested on a hardcoded alsa-rs implementation I hacked together.

I see some blockers for now:

- writing a reliable, crossplatform PCM sound capture code, ideally one that works in a separate thread
- making libcava available on those platforms, so far my FFI bindings only works if there's a pkg-config-detectable installation on the target system, and libcava is far from a release that makes this even remotely possible :/

**Comment by tranzystorek-io** [comment]
Took a look at #14 and symphonia, and if I understand it correctly and we can "tee" the decoded audiosamples to cava, we would be pretty much set, probably just use `dasp_sample` to convert samples on the fly to `f64` that cava understands

**Comment by tranzystorek-io** [comment]
Closing this issue as the problem has been fixed, moving discussion to #15
